### DISCUSSION TITLE
Generic value from JSON to function

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [nlohmann/json#3779] [Q&A]
Hi,

I have a small problem here that im searching a lot to solve.

I have this function:

```
void test(const std::any &param)
{
    auto value = std::any_cast<bool>(param);
    std::cout << "Received Value: " << value << std::endl;
}
```

And i have a generic code the read a JSON, get a function name and the parameter value, that can be bool, i32, or custom class.

```
// main
int main()
{
    // list of mapping
    std::map<std::string, std::function<void(const std::any &)>> flist;
    flist["test"] = [] &MyClass::test;

    // method: test
    {
        // the json come from mobile (kotlin or swift)
        json j = json::parse(R"({
            "function": "test",
            "param": true
        })");

        auto function = j["function"].get<std::string>();
        auto param = j["param"].get<std::any>();
        flist[function](param);
    }
}
```

But nlohmann json don't know `std::any` and i can't use `bool` because only my function `test` that can receive something that it will cast to correct value type. And the function `test` can't receive a nlohmann object or string to be parsed after, because it need receive an independent value to be casted after be called.

The error is:

```
/Users/paulo/Developer/workspaces/cpp/cpp-native-mobile/cxx/src/main.cpp:94:33: error: no matching member function for call to 'get'
        auto param = j["param"].get<std::any>();
                     ~~~~~~~~~~~^~~~~~~~~~~~~
/Users/paulo/Developer/workspaces/cpp/cpp-native-mobile/build/cxx/_deps/json-src/include/nlohmann/json.hpp:1731:10: note: candidate template ignored: substitution failure [with ValueTypeCV = std::any, ValueType = std::any]: no matching member function for call to 'get_impl'
    auto get() const noexcept(
         ^
/Users/paulo/Developer/workspaces/cpp/cpp-native-mobile/build/cxx/_deps/json-src/include/nlohmann/json.hpp:1772:10: note: candidate template ignored: substitution failure [with PointerType = std::any]: no matching member function for call to 'get_ptr'
    auto get() noexcept -> decltype(std::declval<basic_json_t&>().template get_ptr<PointerType>())
         ^               
```

Can anyone help me?

**Comment by falbrechtskirchinger** [comment]
The usual recommendation would be to implement a conversion function for `std::any`. Unfortunately, `std::any` is currently explicitly excluded from being usable with `get()`.
The only option is manually retrieving the value as stored or using a convertible type and converting the result to `std::any`.

**Reply by paulocoutinhox** [reply]
:(

This remove the dynamic part of the project. I need find a way to convert it dynamically. It something is wrong, a try/catch get it, dont have problem.

**Reply by paulocoutinhox** [reply]
I can also make something like this (pseudo code):

```
struct FunctionData
{
    std::function<void(const T &)> func;
    T type;
};

auto a = FunctionData{&MyClass::test, bool};
auto param = j["param"].get<a.type>();
a.func(param);
```

The code is wrong, but i want do something like this.

**Comment by falbrechtskirchinger** [comment]
How about something like this?

```cpp
void test(bool value) {
    std::cout << "Received Value: " << value << std::endl;
}

int main() {
    std::map<std::string, std::function<void(const json &)>> flist;
    flist["test"] = [](const json &j) {
        bool value = j.get<bool>();
        test(value);
    };

    json j{{"function", "test"}, {"param", true}};
    auto function = j["function"].get<std::string>();
    flist[function](j["param"]);
}
```

**Reply by paulocoutinhox** [reply]
This is possible and work, but the problem is this line `bool value = j.get<bool>();`. The type `bool` and `json` need be dynamic or outside this, because it is a convertion layer that transform json into objects. I cant use json object on mapping. 

Need be something like:

1. Mapping receive `function pointer` and the `type`.
2. When some data come, it get the function by name and get the item inside map.
3. Check what `adapter` is used.
4. If is `json` adapter, it try convert the json string that come to the `type` (this is the problematic part) from mapping.

**Reply by falbrechtskirchinger** [reply]
Looks like you need another mapping to convert the JSON value. I don't have enough information to figure this step out for you.
